<a href="https://colab.research.google.com/github/YolandaMDavis/cs109a-fall2020/blob/socio-economic/HousingValuesGeo_NYC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!pip install pygeohash

  Created wheel for pygeohash: filename=pygeohash-1.2.0-py2.py3-none-any.whl size=6162 sha256=7bd1e5b23561b782bccc4d496d0d0cd6d0d1ef9b74ee400c6140662499ee6e9c
  Stored in directory: /root/.cache/pip/wheels/3f/5f/14/989d83a271207dda28232746d63e737a2dbd88ea7f7a9db807
Successfully built pygeohash


In [3]:
%matplotlib inline
import math
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pygeohash as pgh
DATA_ROOT = '/content/drive/My Drive/Final-Project-for-CS109a/'

In [6]:
NYC_prop = pd.read_csv(DATA_ROOT + 'pluto_20v6.csv', usecols=['assesstot', 'latitude', 'longitude'])

In [8]:
NYC_prop.head()

assesstot   latitude  longitude
0      900.0  40.840277 -73.780735
1   805050.0  40.644117 -73.886114
2    27780.0  40.899491 -73.848780
3   962550.0  40.714403 -73.962452
4  6541196.0  40.717140 -73.964803

In [9]:
NYC_prop["GEOHASH"] = [pgh.encode(row["latitude"],row["longitude"]) for idx, row in NYC_prop.iterrows()]

In [10]:
NYC_prop.head()

assesstot   latitude  longitude       GEOHASH
0      900.0  40.840277 -73.780735  dr7834dxb788
1   805050.0  40.644117 -73.886114  dr5rnxjc0u88
2    27780.0  40.899491 -73.848780  dr72xt8by9wx
3   962550.0  40.714403 -73.962452  dr5rsgy4tfy6
4  6541196.0  40.717140 -73.964803  dr5rsuk6x1jd

In [27]:
sum(NYC_prop.GEOHASH.isna())

0

In [11]:
nyc_crime = pd.read_csv(DATA_ROOT + 'covid_analysis/nyc_crime.csv')
nyc_crime.shape

(306656, 36)

In [13]:
nyc_crime.tail()

CMPLNT_NUM  ...                       New Georeferenced Column
306651   453871844  ...   POINT (-73.94255491999998 40.70341571400007)
306652   931476633  ...  POINT (-73.77257302099997 40.595820856000046)
306653   889424078  ...   POINT (-73.88118438299993 40.75600770400007)
306654   951280302  ...  POINT (-73.90913931099993 40.833154599000075)
306655   187190867  ...   POINT (-73.77623390699993 40.67998073800004)

[5 rows x 36 columns]

In [15]:
nyc_crime["GEOHASH"] = [pgh.encode(row["Latitude"], row["Longitude"]) for idx, row in nyc_crime.iterrows()]

In [23]:
nyc_crime.tail()

CMPLNT_NUM  ADDR_PCT_CD  ...       GEOHASH LOCAL_PROP_VALUES
306651   453871844           90  ...  dr5rt3udtddu               NaN
306652   931476633          101  ...  dr5wcq0ry3w8     143857.197802
306653   889424078          115  ...  dr5ryg3kgwkp               NaN
306654   951280302           42  ...  dr72q1mn25xq               NaN
306655   187190867          113  ...  dr5x3jtf14hc     106410.812500

[5 rows x 38 columns]

In [30]:
# define function to apply across the GEOHASH of 
from collections import defaultdict

# setting up home value dict (to 600 meter precision) with geohash. 100 meters was giving too many 0 values (almost 1% of rows)
home_value_dict = defaultdict(list)
for idx, row in NYC_prop.iterrows():
    if pd.notna(row['assesstot']):
      home_value_dict[row['GEOHASH'][:6]].append(row['assesstot'])


In [31]:
home_value_dict['dr5rt3']

[19500.0,
 111840.0,
 247050.0,
 878400.0,
 297900.0,
 182700.0,
 414000.0,
 517050.0,
 66840.0,
 249750.0,
 92999.0,
 94501.0,
 34200.0,
 94920.0,
 349200.0,
 149850.0,
 3421350.0,
 59340.0,
 6300.0,
 450450.0,
 334350.0,
 516600.0,
 86460.0,
 499950.0,
 970650.0,
 237150.0,
 335250.0,
 66300.0,
 790650.0,
 478350.0,
 700650.0,
 74820.0,
 986855.0,
 1591650.0,
 125550.0,
 45960.0,
 66300.0,
 415800.0,
 204750.0,
 597150.0,
 438300.0,
 1174500.0,
 86400.0,
 212400.0,
 70320.0,
 70320.0,
 71520.0,
 371250.0,
 88740.0,
 91440.0,
 125550.0,
 65640.0,
 86460.0,
 73260.0,
 108450.0,
 113400.0,
 91800.0,
 66840.0,
 66720.0,
 1021050.0,
 41700.0,
 167400.0,
 147300.0,
 145800.0,
 581850.0,
 793350.0,
 211950.0,
 66300.0,
 68460.0,
 414900.0,
 543150.0,
 104400.0,
 371250.0,
 494550.0,
 1836900.0,
 162900.0,
 75840.0,
 66300.0,
 278100.0,
 94950.0,
 189900.0,
 711000.0,
 518400.0,
 1567350.0,
 267750.0,
 719100.0,
 451800.0,
 80550.0,
 97650.0,
 50400.0,
 12840.0,
 60660.0,
 13680.0,
 16380.0,

In [32]:
# function for getting values from the dictionary from geohash
from statistics import mean

def get_prop_values(geohash, home_values):
  try:
    mean_property_value = mean(home_values[geohash[:6]])
  except:
    mean_property_value = 0

  return mean_property_value

In [33]:
# prop_values_test = get_prop_values(origin, home_value_dict)
# this should create list of the property values near 
LOCAL_PROP_VALUES = [get_prop_values(row["GEOHASH"], home_value_dict) for idx, row in nyc_crime.iterrows()]

In [34]:
nyc_crime['LOCAL_PROP_VALUES'] = LOCAL_PROP_VALUES

In [37]:
nyc_crime.head(10)

CMPLNT_NUM  ADDR_PCT_CD  ...       GEOHASH LOCAL_PROP_VALUES
0    972326799           81  ...  dr5rmtq3w0gr      2.616263e+05
1    376304873           52  ...  dr72wc598dex      4.867931e+05
2    299326203           75  ...  dr5rr4euvm4j      2.951804e+05
3    674946147          121  ...  dr5r0md8uns2      4.717530e+04
4    416422620          101  ...  dr5wcqycd6g0      1.438572e+05
5    119130081           68  ...  dr5r5shuzyjr      1.663550e+05
6    873514553           41  ...  dr72nwsfw7r6      4.480467e+05
7    383718152           46  ...  dr72qh6st1q1      5.888114e+05
8    990133707           70  ...  dr5rm065jguv      6.223408e+05
9    458329800           73  ...  dr5rmuv8d6by      2.769874e+05
10   485818084           88  ...  dr5rs8tbcvgd      9.011962e+05
11   497826839           42  ...  dr72q22qz559      2.510253e+05
12   843609180           43  ...  dr72r17k3w6x      1.386767e+05
13   827320584           88  ...  dr5rkx28gnbw      1.512527e+06
14   428858939           32  ...  dr72jqxgd619      9.213388e+05
15   622771956           67  ...  dr5rmck07m37      1.560765e+05
16   380792723          113  ...  dr5x3jtf14hc      1.064108e+05
17   220189640           70  ...  dr5rkby8xfj9      9.713200e+05
18   774311943           52  ...  dr72qrw7g197      7.494075e+05
19   918565275           42  ...  dr72q9sntecu      3.043136e+05

[20 rows x 38 columns]

In [42]:
sum(nyc_crime.LOCAL_PROP_VALUES == 0)

224

Only 224 of over 300K rows is this 0. A spot check of one of them shows a very deserted place on Riker's Island, so not surprising!

In [40]:
nyc_crime[nyc_crime.LOCAL_PROP_VALUES == 0].head(10) # 

CMPLNT_NUM  ADDR_PCT_CD  ...       GEOHASH LOCAL_PROP_VALUES
1027    390204599           41  ...  dr72n9z19mg6               0.0
1931    813376489           63  ...  dr5qys01rstt               0.0
3397    736433661           44  ...  dr72n9z19mg6               0.0
4837    333559103            5  ...  dr72n9z19mg6               0.0
7576    266894109           41  ...  dr72n9z19mg6               0.0
10085   759498101           44  ...  dr72n9z19mg6               0.0
12953   905393148           71  ...  dr72n9z19jg5               0.0
17463   833582425           41  ...  dr72n9z19mg6               0.0
18467   815694671            5  ...  dr72n9z19mg6               0.0
19634   396776152            5  ...  dr72n9z19mg6               0.0

[10 rows x 38 columns]

In [43]:
nyc_crime['LOCAL_PROP_VALUES'].describe()

count    3.066560e+05
mean     2.337846e+06
std      1.684878e+07
min      0.000000e+00
25%      1.754780e+05
50%      4.371159e+05
75%      1.365302e+06
max      7.240287e+09
Name: LOCAL_PROP_VALUES, dtype: float64

In [45]:
nyc_crime.to_csv(DATA_ROOT + 'data_prelim/nyc_crime_local_prop_values.csv', index=False)